pkg/signal/main.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package signal
+
+import (
+	"sync"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/utils"
+)
+
+// ErrNilCallback throws if there is no callback function provided
+var ErrNilCallback = errs.Class("must provide a callback function to register")
+
+// Dispatcher is what's loaded onto each service
+type Dispatcher struct {
+	sync.RWMutex
+	source    string
+	callbacks map[string][]Callback
+}
+
+// Callback is the general type for functions in this package
+type Callback func() error
+
+// NewDispatcher is what each service calls to create a dispatch service
+func NewDispatcher(source string) *Dispatcher {
+	return &Dispatcher{
+		source:    source,
+		callbacks: make(map[string][]Callback),
+	}
+}
+
+// Dispatch broadcasts an event across all callbacks
+// If any of them error, it will return that error,
+func (d *Dispatcher) Dispatch(name string) error {
```

Comment:
the string type might actually be bad for the "name" value. the reason is due to namespace collisions. let's say you have two packages, and they both use the same dispatcher, and accidentally both use the `name` "updates". you're going to have some really subtle bugs that you could actually enforce the compiler to check for you!

so there are (at least) two options. option 1: 

this is going to sound straight up bat crazy, but an interface{} may actually be a better choice for a name type here. each package can define it's own type and use a value of that type. because each type is from a different package, they will never be equal, and never show up in the map as the same key. voila, the compiler prevents this entire class of bug.

e.g.:

```
type myDispatchType int

func Whatever() {
  ...
  err = d.Dispatch(myDispatchType(1))
  ...
```

option 2:

another way of dealing with this is to make a GenSym() function (generate symbol) and only accept symbols. like this:

```
type Symbol uint64

var symbolCounter = new(uint64)

func GenSym() Symbol {
  return Symbol(atomic.AddUint64(symbolCounter, 1))
}

func (d *Dispatcher) Dispatch(name Symbol) error {
   ...
```

then each package would do this:

```
var (
  UpdateChannel = signal.GenSym()
)

func Whatever() {
  ...
  err = d.Dispatch(UpdateChannel)
  ...
```